### PR TITLE
[core] update num_returns type annotation to allow for 'streaming'

### PR DESCRIPTION
**Pull request by marwan116**
## Why are these changes needed?

To avoid mypy errors when passing `num_returns="streaming"`

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [x] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [x] Release tests

### CODE REVIEW

File: python/ray/_private/worker.py [hunk 1]
```diff
@@ -3189,7 +3189,7 @@ def remote(__t: type) -> Any:
 @overload
 def remote(
     *,
-    num_returns: Union[int, float] = Undefined,
+    num_returns: Union[int, float, Literal["dynamic"]] = Undefined,
```
Comment (jjyao) [inline python/ray/_private/worker.py:3192]: "dynamic" is replaced by "streaming", could you update here?

Also can you remove `float` since it cannot be float :)

**Review by jjyao — COMMENTED** [review]
Thanks for the contribution and sorry for not reviewing it promptly.

File: python/ray/_private/worker.py (hunk 1 above)
Comment (marwan116) [reply python/ray/_private/worker.py]: should now be addressed

**Comment by marwan116** [comment]
I updated the type annotation in other python modules as well. Perhaps given the typing is shared across modules, it would be good practice to set the num_returns type alias once in a typing module and refactor accordingly....

**Comment by marwan116** [comment]
On thing I spotted is `ActorHandle._actor_method_call` seems to be currently handling `dynamic` and `streaming` differently at the moment - see https://github.com/ray-project/ray/blob/master/python/ray/actor.py#L1420

File: python/ray/util/client/common.py [hunk 2]
```diff
@@ -532,7 +532,7 @@ def __init__(
         self,
         actor_handle: ClientActorHandle,
         method_name: str,
-        num_returns: int,
```
Comment (jjyao) [inline python/ray/util/client/common.py:535]: Ray client doesn't support streaming yet so we can revert this change.

**Review by jjyao — APPROVED** [review]
Thanks for the contribution!! Ping me when the last comment is resolved and I'll merge the PR.

File: python/ray/util/client/common.py (hunk 2 above)
Comment (marwan116) [reply python/ray/util/client/common.py]: should now be addressed
